kratos/utilities/assign_mpcs_to_neighbours_utility.cpp

```diff
@@ -0,0 +1,376 @@
+    if (pNode->HasDofFor(GetComponentVariable(rVariable, 0)) &&
+        pNode->HasDofFor(GetComponentVariable(rVariable, 1)) &&
+        pNode->HasDofFor(GetComponentVariable(rVariable, 2))) {
+
+        rCloud_of_dofs_x.push_back(pNode->pGetDof(GetComponentVariable(rVariable, 0)));
+        rCloud_of_dofs_y.push_back(pNode->pGetDof(GetComponentVariable(rVariable, 1)));
+        rCloud_of_dofs_z.push_back(pNode->pGetDof(GetComponentVariable(rVariable, 2)));
+
+    } else {
+        KRATOS_ERROR << "The node with ID " << pNode->Id() << " does not have the required DOFs for the variable " << rVariable << std::endl;
+    }
+    rSlave_coordinates = pNode->Coordinates();
+    KRATOS_CATCH("");
+}
+
+// Get Dofs and Coordinates arrays for a given variable double. (For nodes)
+void AssignMPCsToNeighboursUtility::GetDofsAndCoordinatesForNodes(
+    ResultNodesContainerType nodes_array,
+    const Variable<double>& rVariable,
+    DofPointerVectorType& rCloud_of_dofs,
+    Matrix& rCloud_of_nodes_coordinates
+    )
+{
+    KRATOS_TRY;
+    rCloud_of_dofs.resize(nodes_array.size());
+    rCloud_of_nodes_coordinates.resize(nodes_array.size(),3);
+    for(unsigned int i = 0; i < nodes_array.size(); i++){
+        rCloud_of_dofs[i] = nodes_array[i]->pGetDof(rVariable);
+        noalias(row(rCloud_of_nodes_coordinates,i)) = nodes_array[i]->Coordinates();
+    }
+    KRATOS_CATCH("");
+}
+
+
+void AssignMPCsToNeighboursUtility::GetDofsAndCoordinatesForNodes(
+    ResultNodesContainerType nodes_array,
+    const Variable<array_1d<double, 3>>& rVariable,
+    DofPointerVectorType& rCloud_of_dofs_x,
+    DofPointerVectorType& rCloud_of_dofs_y,
+    DofPointerVectorType& rCloud_of_dofs_z,
+    Matrix& rCloud_of_nodes_coordinates
+)
+{
+    KRATOS_TRY;
+    rCloud_of_dofs_x.resize(nodes_array.size());
+    rCloud_of_dofs_y.resize(nodes_array.size());
+    rCloud_of_dofs_z.resize(nodes_array.size());
+    rCloud_of_nodes_coordinates.resize(nodes_array.size(), 3);
+
+    for(int i = 0; i < static_cast<int>(nodes_array.size()); i++)
+    {
+        NodeType::Pointer pNode = nodes_array[i];
+
+        // Check if the node has the required DOFs
+        if (pNode->HasDofFor(GetComponentVariable(rVariable, 0)) &&
+            pNode->HasDofFor(GetComponentVariable(rVariable, 1)) &&
+            pNode->HasDofFor(GetComponentVariable(rVariable, 2))) {
+
+            rCloud_of_dofs_x[i] = pNode->pGetDof(GetComponentVariable(rVariable, 0));
+            rCloud_of_dofs_y[i] = pNode->pGetDof(GetComponentVariable(rVariable, 1));
+            rCloud_of_dofs_z[i] = pNode->pGetDof(GetComponentVariable(rVariable, 2));
+
+        } else {
+            KRATOS_ERROR << "The node with ID " << pNode->Id() << " does not have the required DOFs for the variable " << rVariable << std::endl;
+        }
+        noalias(row(rCloud_of_nodes_coordinates, i)) = pNode->Coordinates();
+    }
+    KRATOS_CATCH("");
+}
+
+void AssignMPCsToNeighboursUtility::AssignRotationToNodes(
```

Comment:
This is not used anywhere in this process. Besides, it is a custom function for your target application. Long story short, it must be placed somewhere else.